plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/Simplifier.hs

```diff
@@ -0,0 +1,96 @@
+{-# LANGUAGE NamedFieldPuns #-}
+
+module UntypedPlutusCore.Transform.Simplifier (
+    SimplifierT (..),
+    SimplifierTrace (..),
+    SimplifierStage (..),
+    Simplification (..),
+    runSimplifierT,
+    evalSimplifierT,
+    execSimplifierT,
+    Simplifier,
+    runSimplifier,
+    evalSimplifier,
+    execSimplifier,
+    initSimplifierTrace,
+    recordSimplification,
+) where
+
+import Control.Monad.State (MonadTrans, StateT)
+import Control.Monad.State qualified as State
+
+import Control.Monad.Identity (Identity, runIdentity)
+import PlutusCore.Quote (MonadQuote)
+import UntypedPlutusCore.Core.Type (Term)
+
+newtype SimplifierT name uni fun ann m a =
+  SimplifierT
+    { getSimplifierT :: StateT (SimplifierTrace name uni fun ann) m a
+    }
+  deriving newtype (Functor, Applicative, Monad, MonadTrans)
+
+instance MonadQuote m => MonadQuote (SimplifierT name uni fun ann m)
+
+runSimplifierT
+  :: SimplifierT name uni fun ann m a
+  -> m (a, SimplifierTrace name uni fun ann)
+runSimplifierT = flip State.runStateT initSimplifierTrace . getSimplifierT
+
+evalSimplifierT
+  :: Monad m => SimplifierT name uni fun ann m a -> m a
+evalSimplifierT = flip State.evalStateT initSimplifierTrace . getSimplifierT
+
+execSimplifierT
+  :: Monad m => SimplifierT name uni fun ann m a -> m (SimplifierTrace name uni fun ann)
+execSimplifierT = flip State.execStateT initSimplifierTrace . getSimplifierT
+
+type Simplifier name uni fun ann = SimplifierT name uni fun ann Identity
+
+runSimplifier :: Simplifier name uni fun ann a -> (a, SimplifierTrace name uni fun ann)
+runSimplifier = runIdentity . runSimplifierT
+
+evalSimplifier :: Simplifier name uni fun ann a -> a
+evalSimplifier = runIdentity . evalSimplifierT
+
+execSimplifier :: Simplifier name uni fun ann a -> SimplifierTrace name uni fun ann
+execSimplifier = runIdentity . execSimplifierT
+
+data SimplifierStage
+  = FloatDelay
+  | ForceDelay
+  | CaseOfCase
+  | CaseReduce
+  | Inline
+  | CSE
+
+data Simplification name uni fun a =
+  Simplification
+    { beforeAST :: Term name uni fun a
+    , stage     :: SimplifierStage
+    , afterAST  :: Term name uni fun a
+    }
+
+-- TODO2: we probably don't want this in memory so after MVP
+-- we should consider serializing this to disk
+newtype SimplifierTrace name uni fun a =
+  SimplifierTrace
+    { simplifierTrace
+      :: [Simplification name uni fun a]
+    }
+
+initSimplifierTrace :: SimplifierTrace name uni fun a
+initSimplifierTrace = SimplifierTrace []
+
+recordSimplification
+  :: Monad m
+  => Term name uni fun a
+  -> SimplifierStage
+  -> Term name uni fun a
+  -> SimplifierT name uni fun a m ()
+recordSimplification beforeAST stage afterAST =
+  let simplification = Simplification { beforeAST, stage, afterAST }
+    in
+      modify $ \st ->
+        st { simplifierTrace = simplifierTrace st ++ [simplification] }
```

Comment:
Why not store the list in reversed form? I'm not a fan of easily avoidable quadratic complexity, although maybe it's fine in this case.

Reply:
This is one of those things that is easy to see in a review, but not while writing it :) No one's sanity will cope with this list being reversed in the Agda, and that traverses it "forwards" but we could build it in reverse and reverse it at the last minute before it is passed to Agda?

Reply:
Yes.

Reply:
Oops, thanks for noticing. Yeah, this is a remnant of the "let's get something working quickly" implementation. I'll fix it asap.